### PR TITLE
respect the msize value for read/write

### PR DESCRIPTION
**Pull request by simonferquel**
There was a typo in the msize negociation code, and for large read / writes, the msize value was not taken into account (if the size of the data to write + size of the header message > msize, we must chunk the data)

### CODE REVIEW

**Comment by dgageot** [comment]
ping @stevvooe :-)

File: version.go [hunk 1]
```diff
@@ -38,7 +38,8 @@ func clientnegotiate(ctx context.Context, ch Channel, version string) (string, e
 			return "", fmt.Errorf("unsupported server version: %v", version)
 		}
 
-		if int(v.MSize) > ch.MSize() {
+		if int(v.MSize) < ch.MSize() {
```
Comment (stevvooe) [inline version.go]: Why reverse this condition?

File: client.go [hunk 2]
```diff
@@ -135,39 +135,70 @@ func (c *client) Walk(ctx context.Context, fid Fid, newfid Fid, names ...string)
...
-	})
-	if err != nil {
-		return 0, err
-	}
-
-	rread, ok := resp.(MessageRread)
-	if !ok {
-		return 0, ErrUnexpectedMsg
+	// size[4] Rread tag[2] count[4] data[count]
+	const rreadMessageEnvelopeSize int = 4 + 1 + 2 + 4
```
Comment (stevvooe) [inline client.go]: This calculation should be a part of the message type. Please modify this correctly in codec.

File: client.go [hunk 3]
```diff
@@ -135,39 +135,70 @@ func (c *client) Walk(ctx context.Context, fid Fid, newfid Fid, names ...string)
...
+
+	maxChunkSize := c.msize - rreadMessageEnvelopeSize
+	totalRead := 0
+
+	for {
```
Comment (stevvooe) [inline client.go]: This breaks the abstraction: there should only be a single message sent per method call.

You can create a helper, similar to `io.ReadFull`, to handle chunking.

File: channel.go [hunk 4]
```diff
@@ -172,6 +172,23 @@ func (ch *channel) WriteFcall(ctx context.Context, fcall *Fcall) error {
...
+	// partial read according to msize
+	case MessageTread:
+		emptyReadResponseWithSizePrefix := ch.codec.Size(newFcall(Tag(0), MessageRread{})) + 4
+		overflow := (emptyReadResponseWithSizePrefix + int(m.Count)) - ch.MSize()
+		if overflow > 0 {
+			fcall.Message = MessageTread{Offset: m.Offset, Fid: m.Fid, Count: m.Count - uint32(overflow)}
+		}
+	}
+
 	p, err := ch.codec.Marshal(fcall)
```
Comment (stevvooe) [inline channel.go]: Should this sizing be represented by the codec? Just test `len(p)` against the returned buffer. We can add a `Size` method to codec to avoid alloc.

File: session.go [hunk 5]
```diff
@@ -31,3 +31,30 @@ type Session interface {
...
+			return err
+		}
+		p = p[written:]
+		offset += int64(written)
+	}
+	return nil
+}
+
+// ReadFull reads the whole file and returns it in a slice
+func ReadFull(s Session, ctx context.Context, fid Fid, offset int64) (p []byte, err error) {
```
Comment (stevvooe) [inline session.go]: Go back and look at `io.ReadFull`. It takes a buffer. This is going to cause unnecessary allocs.

File: session.go [hunk 6]
```diff
@@ -31,3 +31,30 @@ type Session interface {
...
+	var buf [4096]byte
```
Comment (stevvooe) [inline session.go]: Right here is the problem: this fixes the buffer size and removes control from the caller.

File: session.go [hunk 7]
```diff
@@ -31,3 +31,30 @@ type Session interface {
 	// session implementation.
 	Version() (msize int, version string)
 }
+
+// WriteFull writes the whole p slice to the specified fid
+func WriteFull(s Session, ctx context.Context, fid Fid, p []byte, offset int64) error {
```
Comment (stevvooe) [inline session.go]: Please order the context argument correctly.

Comment (stevvooe) [inline session.go]: Place these in a separate file. They are just methods to use with session. See `readdir.go` for an example.

File: version.go (hunk 1 above)
Comment (stevvooe) [inline version.go]: The client already sent their size. This scales down the local msize if the server responds with a smaller value. The protocol works like this:

1. Client sends msize.
2. Server responds with msize.
  v.Msize > msize, use server's msize.

At least, this was how I understand from study of 9p. Do you have documentation countering this? What does plan9 do?

Comment (stevvooe) [inline version.go]: Also, please move this to a separate PR.

File: session.go (hunk 5 above)
Comment (stevvooe) [inline session.go]: After some thought here, it would be much better to implement an `io.Reader` or `io.ReadSeeker` that tracks offset and fid. This way, the `io` utility functions can be used composably. The singature would be something like this:

```
NewReadSeeker(ctx context.Context, s Session, fid Fid) (io.ReadSeeker, error)
```

This should open up correct buffering.

**Comment by simonferquel** [comment]
Should be ready to merge. I have successfully tested it in the context of Docker for Windows / Mac.
Some structural changes have been made : 
- Codec has now knowledge of the max size it can use to marshall messages
- This has required some plumbing for tests and 9pr to build and run
- High level clients should prefer to rely on NewFileReader / NewFileWriter that implement io.Reader / io.Writer behaviors to access data streams

This also have a change in the package signature (newly exposed NewFileReader / Writer functions and newly exposed Codec from the session - so that 9pr can be reliable)

**Comment by simonferquel** [comment]
This way of dealing with the problem is too much invasive prefer #29
